### ISSUE TITLE
GCCcore toolchain leads to duplicated modules, confusing lmod

### ISSUE THREAD

**Issue by Micket** [hpcugent/easybuild#322]
I'm asking this question here, because I think the modules produced by EB are a bit strange, even if this might be considered a bug in Lmod.

We use the HMNS, and have found for several modules that that the duplicated modules (M4, Bison, flex, zlib, binutils, and possibly others) that exist in multiple trees cause problems for LMOD.

This leads to cases where some module depend on for example zlib in multiple ways. Just installing a few modules, has led to

    $ ml spider zlib/1.2.8
    ...
    GCC/5.4.0-2.26
    GCC/5.4.0-2.26  OpenMPI/1.10.3
    GCCcore/4.9.3
    GCCcore/5.4.0
    icc/2016.3.210-GCC-5.4.0-2.26  impi/5.1.3.181
    ifort/2016.3.210-GCC-5.4.0-2.26  impi/5.1.3.181
    ...

and this causes strange behavior in Lmod:

    $ ml foss/2016b Tcl/8.6.5
    $ ml OpenMPI/1.10.3
    Lmod has detected the following error: 
    These module(s) exist but cannot be loaded as requested:  
    "zlib/1.2.8"

Which has numerous strange issues;
1. loading zlib/1.2.8 does actually work, 
2. OpenMPI doesn't depend on any zlib, 
3. If I don't load the unrelated "Tcl" module, it works.
4. Loading zlib doesn't change anything.

After all, OpenMPI/1.10.3 is already loaded from foss/2016b.
it's simply Lmod that gets confused when trying to reload modules. In another case the multiple flex-modules caused lmod to get stuck loading/unloading packages until a stack overflow.

Cases like this gets triggered when trying to build other modules (during the sanity check).

Since I don't see everyone else getting these issues, could I have configured EB incorrectly? 
I have simply pointed lmod to the ``all/Core`` path, and as far as I can see, all the modules produced match what the EB-configs state.

Apart from a few paths, the only real option in the EB configuration is
recursive-module-unload=True
which I'm not sure why it was selected, or how it affects the modules (I can't see anything related to unloading).

I've tried different versions of Lmod, including 7.4.

**Comment by boegel** [comment]
You are getting a `zlib/1.2.8` module at different parts of the hierarchy probably because it is a very common dependency, and because we have been including easyconfigs in subtoolchains (e.g. GCCcore) more frequently recently, because EasyBuild v3.x now also considers subtoolchains without using `--minimal-toolchains` (cfr. http://easybuild.readthedocs.io/en/latest/Manipulating_dependencies.html#using-minimal-toolchains-for-dependencies).

Enabling `--recursive-module-unload` changes the default behaviour in which EasyBuild includes an "isloaded" guard around load statements, for example:

```lua
if not isloaded("GCC/5.4.0-2.26") then
    load("GCC/5.4.0-2.26")
end
```

With `--recursive-module-unload` enabled, you'll get this instead:

```
load("GCC/5.4.0-2.26")
```

This means that the `load` will be executed every time the module that includes this is loaded, which results in a so-called "load storm", see also http://lmod.readthedocs.io/en/latest/210_load_storms.html?highlight=load%20storm .

This can be a problem especially in combination with HMNS, since for Lmod a `load` really means loading the module, i.e. it will unload and reload the module every time.

Triggering a reload for a module like `OpenMPI`, which extends `$MODULEPATH` with another part of the hierarchy, can cause other modules to be temporarily unavailable while Lmod is processing loads, leading to the problem you are seeing.

Long story short: I don't think it's a good idea to have `--recursive-module-unload` enabled when using HMNS. The only downside of this (afaik) is that is a user unload a particular module, the modules for the dependencies remain loaded, i.e. the unload isn't recursive as a user may expect.

Thoughts @rtmclay, @ocaisa, @bartoldeman?

**Comment by Micket** [comment]
I copied the whole module tree, and started stripping out component after component trying to reach a minimal set. Turns out that one module was built with filter-deps=zlib, which led to the Tcl(foss) module having no zlib dependency, while Tcl(GCCcore) had it. Adding the missing zlib (making the modules consistent) fixed this issue.
I suspect the other issue I've had with lmod was caused by a similar mismatch of modules.

A huge thanks for your help, it really helped my find the root cause for this particular hickup.

Is there a way to trigger EB to rebuild all the --modules-only for all the installed configs, in case I want to turn off --recursive-module-unload (or just to ensure that no other filter-deps have sneaked into the installation)?

**Comment by boegel** [comment]
@Micket You could try a `--force --module-only` using a copy of all the easyconfigs that are available in the `easybuild` subdirectory in the installation directories, but I recommend backing up the original modules and checking with `diff` afterwards since `--module-only` is known not to be 100% perfect yet (cfr. https://github.com/hpcugent/easybuild-framework/issues/1877).

Alternatively, you can also use something like `--installpath-modules /tmp/test_modules` first to easily diff before you replace the current modules.
